**尝试代码实现一次反向传播

In [31]:
import numpy as np
import math

下面我们来准备一些数据定义

In [32]:
x = np.matrix([[0.5],[1.0]])
t = np.matrix([[1.0],[0.0]])
w1 = np.matrix([[0.1, 0.2],[0.3, 0.4]]).T
w2 = np.matrix([[0.6, 0.7], [0.8, 0.9]]).T
b1 = np.matrix([[0.5], [0.5]])
b2 = np.matrix([[1.0], [1.0]])

进行一次前向计算

In [33]:
print('forward>>>>>>>>')

forward>>>>>>>>


In [34]:
logit1 = np.add(np.dot(w1,x),b1)

In [35]:

def sigmoid(x):
  return 1 / (1 + np.exp(-x))

In [36]:
h1 =sigmoid(logit1)

In [37]:
logit2 = np.add(np.dot(w2,h1),b2)

In [38]:
h2 = sigmoid(logit2)

In [39]:
y=h2

In [40]:
cost = np.sum(np.power((t-y), 2)/2)

In [41]:
print('x:', x)
print('w1:', w1)
print('b1:', b1)
print('w2:', w2)
print('b2:', b2)
print('logit1:', logit1)
print('h1:', h1)
print('logit2:', logit2)
print('h2:', h2)
print('cost:')
print(cost)
print('>>>>>>>>>>end forward')

x: [[0.5]
 [1. ]]
w1: [[0.1 0.3]
 [0.2 0.4]]
b1: [[0.5]
 [0.5]]
w2: [[0.6 0.8]
 [0.7 0.9]]
b2: [[1.]
 [1.]]
logit1: [[0.85]
 [1.  ]]
h1: [[0.70056714]
 [0.73105858]]
logit2: [[2.00518715]
 [2.14834972]]
h2: [[0.88134062]
 [0.89551446]]
cost:
0.40801310159501947
>>>>>>>>>>end forward


In [15]:
y = h2

接下来我们计算一次反向传播

In [44]:
print('<<<<<<<<<<backward')
delta_2 = np.multiply(y-t,np.multiply(y,(1-y)))
nabla_w2 = np.dot(delta_2,h1.T)
nabla_b2 = delta_2
sp_1= np.multiply(h1,(1-h1))
delta_1 =  np.multiply(np.dot(w2.T,delta_2),sp_1)
nabla_w1 = np.dot(delta_1, x.T
nabla_b1 = delta_1
new_w2 = w2 - nabla_w2
new_b2 = b2 - nabla_b2
new_w1 = w1 - nabla_w1
new_b1 = b1 - nabla_b1

print('delta_2', delta_2)
print('sp_1:', sp_1)
print('delta_1', delta_1)
print('nabla_w1', nabla_w1)
print('nabla_b1', nabla_b1)
print('nabla_w2', nabla_w2)
print('nabla_b2', nabla_b2)
print('------------end backward')


<<<<<<<<<<backward
delta_2 [[-0.01240932]
 [ 0.08379177]]
sp_1: [[0.20977282]
 [0.19661193]]
delta_1 [[0.01074218]
 [0.01287516]]
nabla_w1 [[0.00537109 0.01074218]
 [0.00643758 0.01287516]]
nabla_b1 [[0.01074218]
 [0.01287516]]
nabla_w2 [[-0.00869356 -0.00907194]
 [ 0.05870176  0.0612567 ]]
nabla_b2 [[-0.01240932]
 [ 0.08379177]]
------------end backward


```
forward>>>>>>>>
('x:', matrix([[ 0.5],
        [ 1. ]]))
('w1:', matrix([[ 0.1,  0.3],
        [ 0.2,  0.4]]))
('b1:', matrix([[ 0.5],
        [ 0.5]]))
('w2:', matrix([[ 0.6,  0.8],
        [ 0.7,  0.9]]))
('b2:', matrix([[ 1.],
        [ 1.]]))
('logit1:', matrix([[ 0.85],
        [ 1.  ]]))
('h1:', matrix([[ 0.70056714],
        [ 0.73105858]]))
('logit2:', matrix([[ 2.00518715],
        [ 2.14834972]]))
('h2:', matrix([[ 0.88134062],
        [ 0.89551446]]))
cost:
0.408013101595
>>>>>>>>>>end forward
('delta_2', matrix([[-0.01240932],
        [ 0.08379177]]))
('sp_1:', matrix([[ 0.20977282],
        [ 0.19661193]]))
('delta_1', matrix([[ 0.01074218],
        [ 0.01287516]]))
('nabla_w1', matrix([[ 0.00537109,  0.01074218],
        [ 0.00643758,  0.01287516]]))
('nabla_b1', matrix([[ 0.01074218],
        [ 0.01287516]]))
('nabla_w2', matrix([[-0.00869356, -0.00907194],
        [ 0.05870176,  0.0612567 ]]))
('nabla_b2', matrix([[-0.01240932],
        [ 0.08379177]]))